extensions-core/google-extensions/src/main/java/org/apache/druid/data/input/google/GoogleCloudStorageInputSource.java

```diff
@@ -59,23 +65,42 @@ public GoogleCloudStorageInputSource(
   }
 
   @Override
-  protected GoogleCloudStorageEntity createEntity(InputSplit<CloudObjectLocation> split)
+  protected InputEntity createEntity(CloudObjectLocation location)
   {
-    return new GoogleCloudStorageEntity(storage, split.get());
+    return new GoogleCloudStorageEntity(storage, location);
   }
 
   @Override
-  protected Stream<InputSplit<CloudObjectLocation>> getPrefixesSplitStream()
+  protected Stream<InputSplit<List<CloudObjectLocation>>> getPrefixesSplitStream(@Nonnull SplitHintSpec splitHintSpec)
   {
-    return StreamSupport.stream(storageObjectIterable().spliterator(), false)
-                        .map(this::byteSourceFromStorageObject)
-                        .map(InputSplit::new);
+    final Iterator<List<StorageObject>> splitIterator = splitHintSpec.split(
+        storageObjectIterable().iterator(),
+        storageObject -> {
+          final BigInteger sizeInBigInteger = storageObject.getSize();
+          long sizeInLong;
+          if (sizeInBigInteger == null) {
+            sizeInLong = Long.MAX_VALUE;
+          } else {
+            try {
+              sizeInLong = sizeInBigInteger.longValueExact();
+            }
+            catch (ArithmeticException e) {
+              sizeInLong = Long.MAX_VALUE;
```

Comment:
Should this propagate the exception instead? If we get an object with a byte size that can't be stored in a long, something seems very wrong

Reply:
The length of a google storage object is the unsigned long type (https://cloud.google.com/storage/docs/json_api/v1/objects#resource-representations). I think it's better to work instead of failing. Added a warning log about the exception.